Util/ONNXExport/src/main/java/org/tribuo/util/onnx/ONNXOperator.java

```diff
@@ -0,0 +1,228 @@
+     * @param outputs The names of the outputs.
+     * @return The NodeProto.
+     */
+    default public OnnxMl.NodeProto build(ONNXContext context, String input, String[] outputs) {
+        return build(context,new String[]{input},outputs,Collections.emptyMap());
+    }
+
+    /**
+     * Builds this node based on the supplied input and outputs.
+     * Throws {@link IllegalArgumentException} if the number of inputs, outputs or attributes is wrong.
+     * May throw {@link UnsupportedOperationException} if the attribute type is not supported.
+     * @param context The onnx context used to ensure this node has a unique name.
+     * @param input The name of the input.
+     * @param outputs The names of the outputs.
+     * @param attributeValues The attribute names and values.
+     * @return The NodeProto.
+     */
+    default public OnnxMl.NodeProto build(ONNXContext context, String input, String[] outputs, Map<String,Object> attributeValues) {
+        return build(context,new String[]{input},outputs,attributeValues);
+    }
+
+    /**
+     * Builds this node based on the supplied inputs and outputs.
+     * Throws {@link IllegalArgumentException} if the number of inputs or outputs is wrong.
+     * @param context The onnx context used to ensure this node has a unique name.
+     * @param inputs The names of the inputs.
+     * @param outputs The names of the outputs.
+     * @return The NodeProto.
+     */
+    default public OnnxMl.NodeProto build(ONNXContext context, String[] inputs, String[] outputs) {
+        return build(context,inputs,outputs,Collections.emptyMap());
+    }
+
+    /**
+     * Builds this node based on the supplied inputs and outputs.
+     * Throws {@link IllegalArgumentException} if the number of inputs, outputs or attributes is wrong.
+     * May throw {@link UnsupportedOperationException} if the attribute type is not supported.
+     * @param context The onnx context used to ensure this node has a unique name.
+     * @param inputs The names of the inputs.
+     * @param outputs The names of the outputs.
+     * @param attributeValues The attribute names and values.
+     * @return The NodeProto.
+     */
+    default public OnnxMl.NodeProto build(ONNXContext context, String[] inputs, String[] outputs, Map<String,Object> attributeValues) {
+        int numInputs = getNumInputs();
+        int numOptionalInputs = getNumOptionalInputs();
+        int numOutputs = getNumOutputs();
+        String opName = getOpName();
+        String domain = getOpDomain();
+        Map<String, ONNXAttribute> attributes = getAttributes();
+        Set<String> mandatoryAttributeNames = getMandatoryAttributeNames();
+
+        if ((numInputs != VARIADIC_INPUT) && ((inputs.length < numInputs) || (inputs.length > numInputs + numOptionalInputs))) {
+            throw new IllegalArgumentException("Expected " + numInputs + " inputs, with " + numOptionalInputs + " optional inputs, but received " + inputs.length);
```

Comment:
These error messages would be much more helpful in debugging if they contained the `opName`, `domain` and maybe also the expected number of input and output arguments. Perhaps they should all start with something like:
`String.format("For %s:%s(%d(+%d)) -> %d", domain, opName, numInputs numOptionalInputs, numOutputs)`

Reply:
Alternatively, if, from the domain and opName, we can reliably generate a link to the actual ONNX docs for that function, we should probably try to do so.

Reply:
We could generate a link if we recorded the version of that operator, which we don't really have (as the opset only updates some of the ops in each version). So I'd consider that future work. I'll fix the string to include the op name as you suggest.